src/coreclr/hosts/coreshim/CoreShim.cpp

```diff
@@ -0,0 +1,275 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "CoreShim.h"
+
+#include <set>
+#include <sstream>
+#include <vector>
+#include <mutex>
+
+namespace
+{
+    struct PathBuffer
+    {
+        PathBuffer()
+            : DefBuffer{}
+            , Buf{ DefBuffer }
+            , Len{ ARRAYSIZE(DefBuffer) }
+        { }
+
+        void SetLength(_In_ DWORD len)
+        {
+            if (len > Len)
+            {
+                Buf = BigBuffer.data();
+                Len = static_cast<DWORD>(BigBuffer.size());
+            }
+        }
+
+        void ExpandBuffer(_In_ DWORD factor = 2)
+        {
+            SetLength(Len * factor);
+        }
+
+        operator DWORD()
+        {
+            return Len;
+        }
+
+        operator WCHAR *()
+        {
+            return Buf;
+        }
+
+        WCHAR DefBuffer[MAX_PATH];
+        std::vector<WCHAR> BigBuffer;
+
+        WCHAR *Buf;
+        DWORD Len;
+    };
+
+    std::string GetExePath()
+    {
+        PathBuffer buffer;
+        DWORD len = ::GetModuleFileNameW(nullptr, buffer, buffer);
+        while (::GetLastError() == ERROR_INSUFFICIENT_BUFFER)
+        {
+            buffer.ExpandBuffer();
+            len = ::GetModuleFileNameW(nullptr, buffer, buffer);
+        }
+
+        return std::string{ buffer.Buf, buffer.Buf + len };
+    }
+
+    std::wstring GetEnvVar(_In_z_ const WCHAR *env)
+    {
+        DWORD len = ::GetEnvironmentVariableW(env, nullptr, 0);
+        if (len == 0)
+            throw __HRESULT_FROM_WIN32(ERROR_ENVVAR_NOT_FOUND);
+
+        PathBuffer buffer;
+        buffer.SetLength(len);
+        (void)::GetEnvironmentVariableW(env, buffer, buffer);
+
+        return static_cast<WCHAR *>(buffer.Buf);
+    }
+
+    coreclr *s_CoreClrInstance;
+}
+
+HRESULT coreclr::GetCoreClrInstance(_Outptr_ coreclr **instance, _In_opt_z_ const WCHAR *path)
+{
+    if (s_CoreClrInstance != nullptr)
+    {
+        *instance = s_CoreClrInstance;
+        return S_FALSE;
+    }
+
+    try
+    {
+        std::wstring pathLocal;
+        if (path == nullptr)
+        {
+            pathLocal = GetEnvVar(W("CORE_ROOT"));
+        }
+        else
+        {
+            pathLocal = { path };
+        }
+
+        pathLocal.append(W("\\coreclr.dll"));
+
+        AutoModule hmod = ::LoadLibraryExW(pathLocal.c_str() , nullptr, LOAD_WITH_ALTERED_SEARCH_PATH);
```

Comment:
I can change to that, but I want to be clear here this shim is only for the testing environment. There are way bigger issues in here if this was going to be used as the actual product shim.

Reply:
If this is just testing, I'm not concerned. You might still want ```LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR``` to avoid accidentally picking up stray CoreCLRs from your path or working directory though.

Reply:
That is a good suggestion and i will update it. Thanks.